### PR TITLE
perf(Sub*): use direct inheritance for algebraic instances on subtypes

### PR DESCRIPTION
**Pull request by mattrobball**
There are two intertwined changes in this PR: 
1. Remove the uses of `Subtype.coe_injective.toMonoid` and its friends from the hierarchy of algebraic sub-objects. The cost of unifying data in `Function.Injective.toMonoid`, which for example has two `Add` instances as parameters, seems prohibitive. 
2. Extend the wall between `SubmonoidClass` and `Submonoid` through `SubfieldClass` and `Subfield`. Currently in the library `SubmonoidClass` and `Submonoid` are sectioned off from one another. For example, both have a `toMonoid` instance with higher priority on the latter. This is done to avoid generating large typeclass searches for `XClass` when we only care of `Submonoid`'s. Currently, this wall breaks down at `Subsemiring`'s where we call `SubsemiringClass` directly for algebraic instances. 

---

- [ ] there are some more holes in the wall. At the `SubSemigroup` level where everything is `MulMemClass` and at the `SubNonUnitalRing` level where there are only `NonUnitalRingClass` dependent instances. Maybe more. 

Afterwards, it will be a good experiment to adjust priorities of `SubX` classes above `default` to avoid issues like those discussed in this [Zulip topic](https://leanprover.zulipchat.com/#narrow/stream/270676-lean4/topic/instance.20priorities.20and.20parents)

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by mattrobball** [comment]
!bench

**Comment by mattrobball** [comment]
!bench

**Comment by mattrobball** [comment]
!bench

**Comment by mattrobball** [comment]
!bench

**Comment by mattrobball** [comment]
!bench

**Comment by kbuzzard** [comment]
Is `GroupTheory.Transfer` a red herring?

**Review by YaelDillies — COMMENTED** [review]
This diff makes me sad. Do we understand why we can't use `Function.Injective.something` without a big performance penalty?

File: Mathlib/GroupTheory/Subgroup/Basic.lean [hunk 1]
```diff
@@ -766,36 +767,42 @@ theorem mk_eq_one_iff {g : G} {h} : (⟨g, h⟩ : H) = 1 ↔ g = 1 :=
...
+    div_eq_mul_inv := fun _ _ => Subtype.ext (div_eq_mul_inv _ _) }
 #align subgroup.to_group Subgroup.toGroup
 #align add_subgroup.to_add_group AddSubgroup.toAddGroup
 
+-- /-- A subgroup of a group inherits an integer power -/
+-- @[to_additive existing]
+-- instance zpow : Pow H ℤ :=
+--   ⟨fun a n => DivInvMonoid.zpow n a⟩
+-- #align subgroup.has_zpow Subgroup.zpow
+
```
Comment (Ruben-VandeVelde) [inline Mathlib/GroupTheory/Subgroup/Basic.lean:785]: Probably this should be removed?

File: Mathlib/GroupTheory/Submonoid/Operations.lean [hunk 2]
```diff
@@ -4,12 +4,12 @@ Released under Apache 2.0 license as described in the file LICENSE.
 Authors: Johannes Hölzl, Kenny Lau, Johan Commelin, Mario Carneiro, Kevin Buzzard,
 Amelia Livingston, Yury Kudryashov
 -/
-import Mathlib.Algebra.Order.Monoid.Basic
 import Mathlib.Algebra.Order.Ring.Lemmas
 import Mathlib.Algebra.Order.ZeroLEOne
 import Mathlib.GroupTheory.GroupAction.Defs
 import Mathlib.GroupTheory.Submonoid.Basic
 import Mathlib.GroupTheory.Subsemigroup.Operations
+import Mathlib.Algebra.Order.Monoid.Defs
```
Comment (Ruben-VandeVelde) [inline Mathlib/GroupTheory/Submonoid/Operations.lean:12]: Keep the imports sorted please

File: Mathlib/NumberTheory/Padics/Hensel.lean [hunk 3]
```diff
@@ -425,12 +425,13 @@ private theorem soln_unique (z : ℤ_[p]) (hev : F.eval z = 0)
     by_contra fun hne =>
       have : F.derivative.eval soln + q * h = 0 :=
         (eq_zero_or_eq_zero_of_mul_eq_zero this).resolve_right hne
-      have : F.derivative.eval soln = -q * h := by simpa using eq_neg_of_add_eq_zero_left this
+      have : F.derivative.eval soln = -(q * h) := by
+        simpa using eq_neg_of_add_eq_zero_left this
```
Comment (Ruben-VandeVelde) [inline Mathlib/NumberTheory/Padics/Hensel.lean:429]: What's the issue here?

File: Mathlib/RingTheory/NonUnitalSubsemiring/Basic.lean [hunk 4]
```diff
@@ -107,7 +111,26 @@ instance : SetLike (NonUnitalSubsemiring R) R where
...
+  { s.toAddSubmonoid.toAddCommMonoid, s.toSubsemigroup.mul with
+    left_distrib := fun _ _ _ => Subtype.ext <| left_distrib _ _ _
+    right_distrib := fun _ _ _ => Subtype.ext <| right_distrib _ _ _
+    zero_mul := fun _ => Subtype.ext <| zero_mul _
+    mul_zero := fun _ => Subtype.ext <| mul_zero _ }
+
+/-- A non-unital subsemiring of a `NonUnitalSemiring` is a `NonUnitalSemiring`. -/
+instance (priority := 100) toNonUnitalSemiring {R} [NonUnitalSemiring R]
+    (s : NonUnitalSubsemiring R) : NonUnitalSemiring s where
+    mul_assoc := fun _ _ _ => Subtype.ext <| mul_assoc _ _ _
```
Comment (Ruben-VandeVelde) [inline Mathlib/RingTheory/NonUnitalSubsemiring/Basic.lean:126]:
```suggestion
  mul_assoc := fun _ _ _ => Subtype.ext <| mul_assoc _ _ _
```

**Review by Ruben-VandeVelde — COMMENTED** [review]
If the `Function.Injective` definitions are problematic, we should document that

**Comment by YaelDillies** [comment]
Can we have a programmatic way to generate those instances in order to replace the `Function.Injective` declarations?

**Comment by mattrobball** [comment]
> This diff makes me sad. Do we understand why we can't use `Function.Injective.something` without a big performance penalty?

My understanding, built from looking at traces, is that because the `Function.Injective/Surjective` definitions use a ton of data, Lean has to work hard to unify all that data at each call site. This seems to occur repeatedly.

**Comment by mattrobball** [comment]
Another data point: the performance improvements to `Subsemiring.Basic` and `Subring.Basic` from excising the few `Order`ed instances in #10900 is inordinate.

File: Mathlib/RingTheory/NonUnitalSubsemiring/Basic.lean (hunk 4 above)
Comment (mattrobball) [reply Mathlib/RingTheory/NonUnitalSubsemiring/Basic.lean:126]: Done

File: Mathlib/GroupTheory/Submonoid/Operations.lean (hunk 2 above)
Comment (mattrobball) [reply Mathlib/GroupTheory/Submonoid/Operations.lean:12]: Done

File: Mathlib/GroupTheory/Subgroup/Basic.lean (hunk 1 above)
Comment (mattrobball) [reply Mathlib/GroupTheory/Subgroup/Basic.lean:785]: Done

**Comment by eric-wieser** [comment]
To me it seems that the solution is a metaprogram that just unfolds `Function.Injective.foo` before creating the instance.

**Review by eric-wieser — COMMENTED** [review]
I agree with Yael that this diff seems like an aesthetic disaster; it's also a maintainability issue, as it makes it very easy to forget to set `nsmul` or `sub` or similar at the instance site, and end up with diamonds.


It's great to know that there is significant performance to be gained by doing *something* here, but I would strongly rather it not be this.

**Comment by mattrobball** [comment]
> To me it seems that the solution is a metaprogram that just unfolds `Function.Injective.foo` before creating the instance.

I think unfolding is the problem. Providing all the data separately is basically forcing this.

**Comment by YaelDillies** [comment]
What's the "division between `Subring` and `SubringClass`"? I saw you talking about it on Zulip but I understand it.

**Comment by mattrobball** [comment]
The instance graph above something a `Subring` is smaller than above a `Subringclass`. Currently priorities are adjusted lower in the hierarchy to preference say `Submonoid` over `SubmonoidClass`. That stops being the case around `NonUnitalSubsemiring` currently and everything flows into `XClass` resulting in more expensive typeclass synthesis.

**Comment by mattrobball** [comment]
Let me do a spot test with a version of the instance constructors that bundles all the data into parents.

**Comment by mattrobball** [comment]
@YaelDillies @eric-wieser see #11029

**Comment by eric-wieser** [comment]
We should see if #11521 can get the same gains here

**Pull request closed without merging**